docs/design/CONFIG_MANAGEMENT.md

```diff
@@ -0,0 +1,238 @@
+}
+
+// GatewayServer encapsulates all the necessary parameters for starting Envoy Gateway.
+// These can either be set via command line or in the config file.
+type GatewayServer struct {
+	GatewayFlags
+	// TBD Envoy Gateway config CRD.
+	Config v1alpha1.GatewayConfig
+}
+
+// ValidateGatewayServer validates configuration of GatewayServer and returns
+// an error if the input configuration is invalid.
+func (s *GatewayServer) ValidateGatewayServer() error {
+	// Validate flags and config file.
+}
+
+// AddFlags adds flags for a specific GatewayFlags to the specified FlagSet.
+func (f *GatewayFlags) AddFlags(fs *pflag.FlagSet) {
+	// Add command line flags, e.g. config file, enabled services, etc.
+}
+```
+A subset of the Envoy Gateway's configuration parameters may be set by a configuration file, as a substitute for
+command-line flags. Providing parameters via a config file is the recommended approach as it simplifies deployment
+and configuration management. The config file is defined by the `GatewayConfig` struct. The config file must be a YAML
+representation of the parameters in this struct. For example:
+```yaml
+apiVersion: gateway.envoy.io/v1alpha1
+kind: GatewayConfig
+foo:
+  bar: baz
+...
+```
+The `--config` flag specifies the path of the Envoy Gateway configuration file. Envoy Gateway will load its config from
+this file. Command line flags which target the same value as a config file will override that value. If `--config` is
+provided and values are not specified via the command line, the defaults for the config file are applied.
+
+The `serve` command calls `New()` and `Start()` to create and run Envoy Gateway. [Manager][mgr] creates controllers,
+e.g. GatewayClass controller, and provides shared dependencies such as clients, caches, etc. Non-controller
+processes, e.g. xDS server, implement the [Runnable][runnable] interface. This allows Manager to manage these processes
+in the same manner as controller-based processes.
+```go
+package manager
+
+import (
+	"context"
+
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/envoyproxy/gateway/internal/kubernetes/gateway"
+	"github.com/envoyproxy/gateway/internal/kubernetes/gatewayclass"
+	"github.com/envoyproxy/gateway/internal/kubernetes/httproute"
+)
+
+// GatewayManager sets-up dependencies and defines the topology of Envoy Gateway
+// managed components, wiring them together.
+type GatewayManager struct {
+	client  client.Client
+	manager manager.Manager
+	config  *v1alpha1.GatewayConfig
+}
+
+// New creates a new Envoy Gateway using the provided configs.
+// TODO: A Manager requires a rest config to construct a clutser. This means
+// using Manager to manage services requires access to a k8s cluster.
+// xref: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.12.1/pkg/cluster/cluster.go#L145-L205
```

Comment:
Recording some notes from the discussion in the community meeting:
- @danehans is going to look into another framework, and make sure it has Context support.
- the rest of the PR is reviewable already.

Reply:
From the call:

So it sounds like we're thinking that we'll need to have two goroutine managers: the controller-runtime one and a non-kube one, possibly nested.  IMO, https://pkg.go.dev/github.com/datawire/dlib/dgroup is attractive for this because it takes a `func(ctx context.Context) error` function pointer as the argument, which happens to be the same signature as controller-runtime's [`manager.Runnable`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.1/pkg/manager#Runnable)'s `Start` method; instead of saying `manager.Add(myRunnable)`, you'd say `group.Go("name", myRunnable.Start)`.

Reply:
I would prefer using https://github.com/tetratelabs/run which also provides the ability to use context for life cycle management of go routines. It provides a superset of functionality compared to `dgroup` and specifically also provides a notion of readiness through different stages of a routine lifecycle, which `dgroup` omits. This due to it being more than just go routine management but even more so a reusable component manager. It allows components to have their own configuration management based on https://github.com/spf13/pflag and provides a dedicated deterministically ordered pre-run wiring and set-up phase next to routine start/stop.

Based on it we (at Tetrate) have a bunch of components we can easily reuse and have configured in a consistent manner (e.g. gRPC server. HTTP server, Signal handler, etc.) Due to strong guarantees on lifecycle management it makes things as registering add-on gRPC services to a gRPC server easy and straight forward. Given we want EG to be highly extensible and easy to wrap/embed by other projects or products I think it would be the right choice.

Reply:
@basvanbeek thank you for providing additional context for https://github.com/tetratelabs/run. Since https://github.com/tetratelabs/run and https://pkg.go.dev/github.com/datawire/dlib/dgroup 
 are based on https://github.com/oklog/run/, maybe we simply proceed with the upstream?

@arkodg @LukeShu @alexgervais @youngnick @skriss please provide input so I can proceed with iterating on the design.

Reply:
I would prefer picking one of our two more complete options than going with the upstream `oklog/run` and having to re-implement parts of functionalities we'll require anyway. I'm familiar with `datawire/dlib/dgroup` so it would be my go-to choice, but that's only because I haven't worked with `tetratelabs/run`.

Reply:
To clarify, `dgroup` is not built on `oklog/run`. It is built on a forked/vendored version of [`golang.org/x/sync/errgroup`](https://pkg.go.dev/golang.org/x/sync/errgroup).

Reply:
@danehans I have very briefly considered to upstream the logic of `tetratelabs/run` when I started building it. 

To be honest, `oklog/run` is a finished library. It serves the scope that @peterbourgon envisioned (goroutine lifecycle management prior to context.Context cancellation being available as standard lib) and is stable (not receiving any updates, doesn't have issues, etc.). The library is very small by itself so it carries more an idea than code.

`tetratelabs/run`'s scope is much bigger. It has native support for context cancellation as well as services that have start/stop semantics. It provides more a framework for reusable components that allow for easy composition and self contained configuration management, bootstrap logic and allows ordered initialization and pre-run phases.

I agree with @alexgervais that choosing native `oklog/run` is a bad idea. However, as I expressed above, I would prefer `tetratelabs/run` over `dgroup` specifically for its inclusion of deterministically ordered configuration and pre goroutine run stages (things specifically mentioned to not be dealt with by `dgroup`. 

The reason I request for these additional features is that it allows the component authors to have consistent handling and management of their components and isolated from others, while keeping a binary's wiring small and clean. This allows projects and Vendor products to reuse as much as possible, having an easy way to embed/wrap EG in their code without having to deal with untangling EG bootstrap or having to duplicate lots of code that will need constant sync with upstream.